{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding Hidden Message in DNA\n",
    "## Week 3\n",
    "\n",
    "**Randomized Algorithms** <br>\n",
    "Will use random decisions to search for motifs.\n",
    "* Las Vegas Algorithms - deliver solutions guaranteed to be exact, despite reliance on random decisions\n",
    "* Monte Carlo Algorithms - Quickly find approximate solutions, not guaranteed to be correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper Functions from previous weeks\n",
    "def symToNum(symbol):\n",
    "    symMap = {'A': 0,\n",
    "              'C': 1,\n",
    "              'G': 2,\n",
    "              'T': 3}\n",
    "    return symMap[symbol]\n",
    "\n",
    "def numToSym(num):\n",
    "    numMap = {0: 'A',\n",
    "              1: 'C',\n",
    "              2: 'G',\n",
    "              3: 'T'}\n",
    "    return numMap[num]\n",
    "\n",
    "def numToPat(index, k):\n",
    "    if k == 1:\n",
    "        return numToSym(index)\n",
    "    pfixIndex = int(index / 4)\n",
    "    r = index % 4\n",
    "    sym = numToSym(r)\n",
    "    pfixPattern = numToPat(pfixIndex, k - 1)\n",
    "    return pfixPattern + sym\n",
    "\n",
    "def patToNum(pattern):\n",
    "    if not pattern:\n",
    "        return 0\n",
    "    symbol = pattern[-1]\n",
    "    prefix = pattern[:-1]\n",
    "    return 4 * patToNum(prefix) + symToNum(symbol)\n",
    "\n",
    "def hammDist(seq1, seq2, d=-1):\n",
    "    dist = 0\n",
    "    if len(seq1) != len(seq2):\n",
    "        return -1\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            dist += 1\n",
    "        if d != -1 and dist > d:\n",
    "            return -1\n",
    "    return dist\n",
    "\n",
    "def profileMost(Text, k, Profile):\n",
    "    profMax = 0.0\n",
    "    kmerMax = Text[:k]\n",
    "    for i in range(len(Text) - k + 1):\n",
    "        kmer = Text[i:i+k]\n",
    "        tProf = 1.0\n",
    "        for x in range(k):\n",
    "            tProf *= Profile[symToNum(kmer[x])][x]\n",
    "        if tProf > profMax:\n",
    "            profMax = tProf\n",
    "            kmerMax = kmer\n",
    "    return kmerMax\n",
    "\n",
    "def makeProfile(Dna, pCounts=True):\n",
    "    prof = [[0] * len(Dna[0]) for _ in range(4)]\n",
    "    for string in Dna:\n",
    "        for i in range(len(string)):\n",
    "            prof[symToNum(string[i])][i] += 1\n",
    "    if pCounts:\n",
    "        return [[(x+1)/(len(Dna) + 4) for x in i] for i in prof]\n",
    "    else:\n",
    "        return [[x/t for x in i] for i in prof]\n",
    "\n",
    "def makeCon(prof):\n",
    "    con = ''\n",
    "    test = list(zip(*prof))\n",
    "    for pos in test:\n",
    "        con += numToSym(pos.index(max(pos)))\n",
    "    return con\n",
    "\n",
    "def score(motifs, pCounts=True):\n",
    "    prof = makeProfile(motifs, pCounts)\n",
    "    con = makeCon(prof)\n",
    "    score = [hammDist(con, x) for x in motifs]\n",
    "    return sum(score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Randomized Motif Search\n",
    "\n",
    "1. We utilize a function that generates a random selection of k-mers (motif) from our Dna strings\n",
    "2. We calculate the profile of this motif\n",
    "3. We generate a new motif, for each string in Dna we calculate the profile most probable k-mer\n",
    "4. We recalculate the profile & score and continue iterating as long as the score continues to decrease\n",
    "\n",
    "We expect the profile of a randomly selected motif to be a uniform profile (all probabilities 1/4) compared to the opposite which is randomly choosing the implanted k-mers correctly the first time resulting in our optimal profile. In practice we obtain a profile somewhere in the middle. Choosing a random motif and iterating allows us to direct the profile away from the uniform toward the ideal profile based off the implanted motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CGATAA',\n",
       " 'GGTTAA',\n",
       " 'GGTATA',\n",
       " 'GGTTAA',\n",
       " 'GGTTAC',\n",
       " 'GGTTAA',\n",
       " 'GGCCAA',\n",
       " 'GGTTAA']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Randomized Motif Search\n",
    "import random as rd\n",
    "\n",
    "def randMotif(Dna, k):\n",
    "    rd.seed()\n",
    "    ran = rd.randint(0,len(Dna[0]) - k)\n",
    "    res = []\n",
    "    for seq in Dna:\n",
    "        res.append(seq[ran:ran+k])\n",
    "    return res\n",
    "\n",
    "def mostProb(tProf, Dna, k):\n",
    "    newMotif = []\n",
    "    for seq in Dna:\n",
    "        newMotif.append(profileMost(seq, k, tProf))\n",
    "    return newMotif\n",
    "\n",
    "def randMotifSearch(Dna, k, t):\n",
    "    gMotif = randMotif(Dna, k)\n",
    "    gScore = score(gMotif)\n",
    "    tMotif = gMotif\n",
    "    while True:\n",
    "        tProf = makeProfile(tMotif)\n",
    "        tMotif = mostProb(tProf, Dna, k)\n",
    "        tScore = score(tMotif)\n",
    "        if tScore < gScore:\n",
    "            gScore = tScore\n",
    "            gMotif = tMotif\n",
    "        else:\n",
    "            return gMotif\n",
    "\n",
    "def randMotifDrive(Dna, k, t, N):\n",
    "    bestMotif = randMotifSearch(Dna, k, t)\n",
    "    bestScore = score(bestMotif)\n",
    "    for _ in range(N):\n",
    "            tMotif = randMotifSearch(Dna, k, t)\n",
    "            tScore = score(tMotif)\n",
    "            if tScore < bestScore:\n",
    "                bestScore = tScore\n",
    "                bestMotif = tMotif\n",
    "    return bestMotif\n",
    "\n",
    "# Failing Debug Tests\n",
    "#6, 8\n",
    "#dna = ['GCACATCATTAAACGATTCGCCGCATTGCCTCGATTAACC', 'TCATAACTGACACCTGCTCTGGCACCGCTCATCCAAGGCC', 'AAGCGGGTATAGCCAGATAGTGCCAATAATTTCCTTAACC', 'AGTCGGTGGTGAAGTGTGGGTTATGGGGAAAGGCAAGGCC', 'AACCGGACGGCAACTACGGTTACAACGCAGCAAGTTAACC', 'AGGCGTCTGTTGTTGCTAACACCGTTAAGCGACGAAGGCC', 'AAGCTTCCAACATCGTCTTGGCATCTCGGTGTGTTTAACC', 'AATTGAACATCTTACTCTTTTCGCTTTCAAAAAAAAGGCC']\n",
    "\n",
    "#8, 5\n",
    "#dna = ['CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA', 'GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG', 'TAGTACCGAGACCGAAAGAAGTATACAGGCGT', 'TAGATCAAGTTTCAGGTGCACGTCGGTGAACC', 'AATCCACCAGCTCCACGTGCAATGTTGGCCTA']\n",
    "\n",
    "#Working Debug Tests\n",
    "## Beginning Off-by-one error test\n",
    "# 6,8\n",
    "dna = ['AATTGGCACATCATTATCGATAACGATTCGCCGCATTGCC', 'GGTTAACATCGAATAACTGACACCTGCTCTGGCACCGCTC', 'AATTGGCGGCGGTATAGCCAGATAGTGCCAATAATTTCCT', 'GGTTAATGGTGAAGTGTGGGTTATGGGGAAAGGCAGACTG', 'AATTGGACGGCAACTACGGTTACAACGCAGCAAGAATATT',  'GGTTAACTGTTGTTGCTAACACCGTTAAGCGACGGCAACT', 'AATTGGCCAACGTAGGCGCGGCTTGGCATCTCGGTGTGTG', 'GGTTAAAAGGCGCATCTTACTCTTTTCGCTTTCAAAAAAA']\n",
    "randMotifDrive(dna, 6, 8, 1000)\n",
    "\n",
    "# with open(\"./Data/dataset_161_5.txt\") as inFile:\n",
    "#     data = inFile.readlines()\n",
    "    \n",
    "# args = data[0].strip().split(' ')\n",
    "# k = int(args[0])\n",
    "# t = int(args[1])\n",
    "# dna = []\n",
    "# for i in range(1, len(data)):\n",
    "#     dna.append(data[i].strip())\n",
    "# print(*randMotifDrive(dna, k ,t),sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "586\n",
      "0.9982935153583617\n",
      "0.9830656030708897\n",
      "0.01693439692911025\n"
     ]
    }
   ],
   "source": [
    "#Compute the probability that ten randomly selected 15-mers from the ten 600-nucleotide long sequences\n",
    "# capture at least one implanted 15-mer.\n",
    "\n",
    "#P(at least 1 implanted k-mer) = 1 - P(0 implanted k-mers)\n",
    "\n",
    "#P(0 implanted k-mers in 10 strings)\n",
    "print(600 - 15 + 1)\n",
    "# = 586\n",
    "#so out of 586 possible k-mer start locations only 1 is correct\n",
    "\n",
    "# For one string the probability of choosing incorrect k-mer is \n",
    "print(585/586)\n",
    "# = 0.9983\n",
    "\n",
    "# For 10 strings this is\n",
    "print((585/586) ** 10)\n",
    "# = 0.9831\n",
    "#So our answer is 1 minus that probability\n",
    "print(1-((585/586) ** 10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.001683326593321766\n",
      "0.00012985670567623384\n"
     ]
    }
   ],
   "source": [
    "#Compute the probability that ten randomly selected 15-mers from ten 600-nucleotide long strings\n",
    "# capture at least two implanted 15-mers.\n",
    "\n",
    "# P(at least two implanted k-mer) = 1 - P(0 implanted k-mers) - P(1 implanted k-mer)\n",
    "print((1/585) * ((585/586) ** 9))\n",
    "print(1 - ((600-15)/(600-15+1))**10 - 10*(600-15)**9/(600-15+1)**10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our randomized motif search all t strings in motif may change in a single iteration. This strategy might be potentially discarded. Utilizing a GibbsSampler, a more cautious approach that discards a single k-mer from the current set of motifs at each iteration.\n",
    "\n",
    "To describe how a GibbsSampler updates Motifs, we will need a more advanced random number generator. Given a probability distribution the number generator models an n-sided die and returns integer i with probability $p_{i}$<br>\n",
    "Gibbs Outline:\n",
    "1. Form <em>Motifs</em> by randomly selecting a k-mer from each sequence\n",
    "2. Randomly choose one of the selected k-mers and remove from <em>Motifs</em>\n",
    "3. Create Profile from the remaining k-mers in <em>Motifs</em>\n",
    "4. For each k-mer that was removed calculate \n",
    "\n",
    "We have defined a Profile-most probable k-mer in a string, now we define a Profile-randomly generated k-mer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCATTTGGAAGGCGT\n",
      "ATAGCCGGTCGGGTT\n",
      "AGGGATGGTCGGGTT\n",
      "ATATATCAACGGGTT\n",
      "ATATATGGTGTTGTT\n",
      "ATATATGAATGGGTT\n",
      "ATATATTTCCGGGTT\n",
      "GCGTATGGTCGGGTT\n",
      "ATCGTTGGTCGGGTT\n",
      "ATATATGGTCGGACG\n",
      "ATATATGGTCCTCTT\n",
      "ATATATGGGAAGGTT\n",
      "ATAGGGGGTCGGGTT\n",
      "GAATATGGTCGGGTC\n",
      "ATATATGGTCGCCAT\n",
      "CTATATGGTCGGGCC\n",
      "ATATCATGTCGGGTT\n",
      "ATATTGAGTCGGGTT\n",
      "ATATAGATTCGGGTT\n",
      "ATATACATTCGGGTT\n",
      "Wall time: 1min 40s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Gibbs Sampler\n",
    "\n",
    "def profileRand(Text, k, tProf):\n",
    "    probs = []\n",
    "    for i in range(len(Text) - k + 1):\n",
    "        tProb = 1.0\n",
    "        kmer = Text[i:i+k]\n",
    "        for i, base in enumerate(kmer):\n",
    "            tProb *= tProf[symToNum(base)][i]\n",
    "        probs.append(tProb)\n",
    "    probSum = sum(probs)\n",
    "    probs = [x/probSum for x in probs]\n",
    "    choice = rd.choices([x for x in range(len(probs))], weights=probs, k=1)\n",
    "    return Text[choice[0]:choice[0]+k]\n",
    "\n",
    "def gibbs(Dna, k, t, N):\n",
    "    motifs = randMotif(Dna, k)\n",
    "    bestMotif = motifs\n",
    "    bestScore = score(bestMotif)\n",
    "    for _ in range(N):\n",
    "        i = rd.randint(0,t-1)\n",
    "        tProf = makeProfile(motifs[:i] + motifs[i+1:])\n",
    "        motifs[i] = profileRand(Dna[i], k, tProf)\n",
    "        tScore = score(motifs)\n",
    "        if tScore < bestScore:\n",
    "            bestScore = tScore\n",
    "            bestMotif = motifs[:]\n",
    "    return bestMotif\n",
    "\n",
    "def gibbsDrive(Dna, k, t, N):\n",
    "    bestMotif = gibbs(Dna, k, t, N)\n",
    "    bestScore = score(bestMotif)\n",
    "    for _ in range(30):\n",
    "        tMotif = gibbs(Dna, k, t, N)\n",
    "        tScore = score(tMotif)\n",
    "        if tScore < bestScore:\n",
    "            bestScore = tScore\n",
    "            bestMotif = tMotif\n",
    "    return bestMotif\n",
    "\n",
    "def gibbsDriveWithRandomMotif(Dna, k, t, N):\n",
    "    bestMotif = randMotifDrive(Dna, k, t, int(N/4))\n",
    "    bestScore = score(bestMotif)\n",
    "    for _ in range(30):\n",
    "        tMotif = gibbs(Dna, k, t, N)\n",
    "        tScore = score(tMotif)\n",
    "        if tScore < bestScore:\n",
    "            bestScore = tScore\n",
    "            bestMotif = tMotif\n",
    "    return bestMotif\n",
    "\n",
    "# dna = ['CGCCCCTCTCGGGGGTGTTCAGTAACCGGCCA',\n",
    "#        'GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG',\n",
    "#        'TAGTACCGAGACCGAAAGAAGTATACAGGCGT',\n",
    "#        'TAGATCAAGTTTCAGGTGCACGTCGGTGAACC',\n",
    "#        'AATCCACCAGCTCCACGTGCAATGTTGGCCTA']\n",
    "# print(gibbsDrive(dna, 8, 5, 100))\n",
    "# print(gibbsDriveWithRandomMotif(dna, 8, 5, 100))\n",
    "# print(gibbsDriveImprove(dna, 8, 5, 100))\n",
    "\n",
    "\n",
    "with open(\"./Data/dataset_163_4.txt\") as inFile:\n",
    "    data = inFile.readlines()\n",
    "\n",
    "args = data[0].strip().split(' ')\n",
    "k = int(args[0])\n",
    "t = int(args[1])\n",
    "N = int(args[2])\n",
    "dna = []\n",
    "for i in range(1, len(data)):\n",
    "    dna.append(data[i].strip())\n",
    "print(*gibbsDrive(dna, k, t, N), sep=\"\\n\")\n",
    "# print(*gibbsDriveWithRandomMotif(dna, k, t, N), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCA\n",
      "CCT\n",
      "CTT\n",
      "TTT\n"
     ]
    }
   ],
   "source": [
    "tProf = makeProfile(['CCA', 'CCT', 'CTT', 'TTG'])\n",
    "dna = ['AAGCCAAA',\n",
    "       'AATCCTGG',\n",
    "       'GCTACTTG',\n",
    "       'ATGTTTTG']\n",
    "for seq in dna:\n",
    "    print(profileMost(seq, 3, tProf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RandomMotifSearch consesus seq: GGGGACTTACGACCCTAGCC Score: 62\n",
      "\n",
      "Gibbs consesus seq: GGGGACCTACGTCCCTAGCC Score: 55\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(\"./Data/DosR.txt\") as inFile:\n",
    "    data = inFile.readlines()\n",
    "\n",
    "dna = []\n",
    "for string in data:\n",
    "    dna.append(string.strip())\n",
    "\n",
    "# randMotifSrch = dict()\n",
    "# for k in range(8, 13):\n",
    "#     tMotif = randMotifDrive(dna, k, len(dna), 50)\n",
    "#     tScore = score(tMotif)\n",
    "#     randMotifSrch[k] = (makeCon(makeProfile(tMotif)), tScore)\n",
    "# print(randMotifSrch)\n",
    "\n",
    "tMotif = randMotifDrive(dna, 20, len(dna), 75)\n",
    "tProf = makeProfile(tMotif)\n",
    "con1 = makeCon(tProf)\n",
    "print(f\"RandomMotifSearch consesus seq: {con1} Score: {score(tMotif)}\\n\")\n",
    "\n",
    "tMotif = gibbsDrive(dna, 20, len(dna), 200)\n",
    "tProf = makeProfile(tMotif)\n",
    "con2 = makeCon(tProf)\n",
    "print(f\"Gibbs consesus seq: {con2} Score: {score(tMotif)}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
