{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding Hidden Message in DNA\n",
    "## Week 3\n",
    "\n",
    "**\"Clock\" Gene**\n",
    "\n",
    "Circadian clock controls daily schedules of animals, plants, and even bacteria. Today many genes have been discovered that play a role in regulating circadian rhythm.\n",
    "\n",
    "First we will focus on plants, as circadian clock is a matter of life or death. Estimates on plant circadian genes are over a thousand, related to photosynthesis, photo reception, and flowering. Every plant cell keeps track of day/night independently of other cells. Genes LHY, CCA1, and TOC1 are the master time keeping regulatory genes and the regulatory proteins they encode are often controlled by external factors (nutrients, sunlight, etc). These regulatory proteins are **transcription factors** that inhibit or promotes expression of genes. A transcription factor bind to a specific short DNA interval called a **regulatory motif or transcription factor binding site** in the genes upstream region (600-1000bp).\n",
    "\n",
    "Since these regulatory motifs are not completely conserved how can we locate them without knowing what they look like in advance?\n",
    "* We could utilize the Frequent word with mismatches algorithm used in week 2, however this was aimed at shorter k-mers with fewer mismatches. Plus finding a DnaA box which is a pattern that appears frequently in a relatively short interval of the genome, whereas a regulatory motif is a pattern that appears at least once (with possible variation) in many different regions scattered throughout the genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper Functions from previous weeks\n",
    "def symToNum(symbol):\n",
    "    symMap = {'A': 0,\n",
    "              'C': 1,\n",
    "              'G': 2,\n",
    "              'T': 3}\n",
    "    return symMap[symbol]\n",
    "\n",
    "def numToSym(num):\n",
    "    numMap = {0: 'A',\n",
    "              1: 'C',\n",
    "              2: 'G',\n",
    "              3: 'T'}\n",
    "    return numMap[num]\n",
    "\n",
    "def numToPat(index, k):\n",
    "    if k == 1:\n",
    "        return numToSym(index)\n",
    "    pfixIndex = int(index / 4)\n",
    "    r = index % 4\n",
    "    sym = numToSym(r)\n",
    "    pfixPattern = numToPat(pfixIndex, k - 1)\n",
    "    return pfixPattern + sym\n",
    "\n",
    "def patToNum(pattern):\n",
    "    if not pattern:\n",
    "        return 0\n",
    "    symbol = pattern[-1]\n",
    "    prefix = pattern[:-1]\n",
    "    return 4 * patToNum(prefix) + symToNum(symbol)\n",
    "\n",
    "def hammDist(seq1, seq2, d=-1):\n",
    "    dist = 0\n",
    "    if len(seq1) != len(seq2):\n",
    "        return -1\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            dist += 1\n",
    "        if d != -1 and dist > d:\n",
    "            return -1\n",
    "    return dist\n",
    "\n",
    "def neighbors(pattern, d):\n",
    "    nucleos = {'A', 'C', 'G', 'T'}\n",
    "    if d == 0:\n",
    "        return pattern\n",
    "    if len(pattern) == 1:\n",
    "        return nucleos\n",
    "    hood = set()\n",
    "    sufNeighbors = neighbors(pattern[1:], d)\n",
    "    for neighbor in sufNeighbors:\n",
    "        if hammDist(pattern[1:], neighbor) < d:\n",
    "            for nucleotide in nucleos:\n",
    "                hood.add(nucleotide + neighbor)\n",
    "        else:\n",
    "            hood.add(pattern[:1] + neighbor)\n",
    "    return hood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motifEnumeration(Dna, k, d):\n",
    "    '''\n",
    "    Given a collection of strings, integers k, and d. Find all motifs (kmers) that are k length and\n",
    "        appear in all strings within the collection with no more than d mismatches. Brute-Force Method.\n",
    "        \n",
    "    Args:\n",
    "        Dna (list[string]): Collection of DNA sequences\n",
    "        k (int): Length of motif(s) we would like to find\n",
    "        d (int): Max mismatches (distance) between k-mer and matches.\n",
    "        \n",
    "    Returns:\n",
    "        list[string]: all k-mers (motif(s)) that appear in each sequence within the Dna collection\n",
    "            allowing no more than d mismatches.\n",
    "    '''\n",
    "    patterns = set()\n",
    "    for i in range(len(Dna[0]) - k + 1):\n",
    "        kmer = Dna[0][i:i+k]\n",
    "        hood = neighbors(kmer, d)\n",
    "        for candidate in hood:\n",
    "            check = 0\n",
    "            for seq in Dna:\n",
    "                check = 0\n",
    "                for i in range(len(seq) - k + 1):\n",
    "                    tk = seq[i:i+k]\n",
    "                    if hammDist(tk, candidate, d) > -1:\n",
    "                        check = 1\n",
    "                        break\n",
    "                if check != 1:\n",
    "                    break\n",
    "            if check:\n",
    "                patterns.add(candidate)\n",
    "    return list(patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGCCG GGCTG AGGCG CGCGG CGGCG GGAGG TGGCG GGCAG GGCGG GCCGG GGCGC GGGCG\n"
     ]
    }
   ],
   "source": [
    "# test/run motifEnumeration function\n",
    "\n",
    "# test = ['ATTTGGC', 'TGCCTTA', 'CGGTATC', 'GAAAATT']\n",
    "# print(motifEnumeration(test, 3, 1))\n",
    "\n",
    "with open(\"./Data/dataset_156_8.txt\") as inFile:\n",
    "    data = inFile.readlines()\n",
    "\n",
    "params = data[0].strip().split(' ')\n",
    "dna = []\n",
    "for i in range(1, len(data)):\n",
    "    dna.append(data[i].strip())\n",
    "\n",
    "print(*motifEnumeration(dna, int(params[0]), int(params[1])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Think about the run time of the brute force algorithm:\n",
    "1. Runs through each kmer of the first DNA seq in the collection (len(seq) - k + 1)\n",
    "2. Calculates all neighbors for the kmer\n",
    "3. For each neighbor, examine all kmers of each subsequent DNA seq in the collection and see if there is a match within d distance.\n",
    "\n",
    "$O(n^t \\times k \\times t)$ Needless to say this algorithm would not scale well at all for larger values of k or d\n",
    "\n",
    "What can we do differently?\n",
    "\n",
    "**Scoring Motifs**\n",
    "\n",
    "* Consider a list of t DNA strings(Dna), where each string has length n, and select a k-mer from each string to form a collection Motifs.\n",
    "\n",
    "\n",
    "1. Score(Motifs) => # of unpopular (non-consensus) letters in a column of the motif matrix\n",
    "2. Count(Motifs) => Construct the 4 x k count matrix. Count the number of occurrences of each nucleotide in each column of the motif matrix\n",
    "3. Profile(Motifs) => Divide each element of the count matrix by t\n",
    "4. Generate a consensus string from the most popular letter in each column of the motif matrix\n",
    "\n",
    "<img src=\"img/motifs_score_count_profile_consensus.png\" width=\"500\">\n",
    "\n",
    "How is conservation accounted for? Think about column 2 (6C, 2A, 2T) and column 12 (6C, and 4T). They both contribute 4 to Score, does this make sense biologically?\n",
    "* No, nucleotides with frequencies exceeding 0.4 should be weighted differently. In the example column 12 is \"more conserved\"\n",
    "\n",
    "Every column of the profile matrix corresponds to a probability distribution (collection of non-negative numbers that sum to 1).\n",
    "\n",
    "**Entropy** is a measure of uncertainty of a probability distribution\n",
    "\n",
    "**$H = -\\sum \\limits_{i=1}^{N} p_{i} * log_{2}p_{i}$**\n",
    "\n",
    "Example:\n",
    "* $log_{2}(0)$ is undefined but multiplied by 0 we assume is 0\n",
    "* Row 2 = $-(0.2log_{2}(0.2) + 0.6log_{2}(0.6) + 0.0log_{2}(0.0) + 0.2log_{2}(0.2)) =1.371$\n",
    "* Row 12 = $-(0.0log_{2}(0.0) + 0.6log_{2}(0.6) + 0.0log_{2}(0.0) + 0.4log_{2}(0.4)) =0.971$\n",
    "\n",
    "Minimum Entropy = 0, where all nucleotides are completely conserved\n",
    "Maximum Entropy = 2, where all probabilities are 1/4\n",
    "\n",
    "The more conserved = the smaller the entropy\n",
    "\n",
    "Instead of exploring all Motifs in Dna and deriving the consensus string from Motifs afterwards. We can explore all potential k-mer consensus strings first and then find the best possible collection Motifs for each consensus string.\n",
    "\n",
    "**Median String Problem**\n",
    "\n",
    "Double minimization problem, we must find a k-mer Pattern that minimizes d(Pattern, Dna) where this function itself is computed by taking a minimum over all choices of k-mers from each string in Dna.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distPatString(Pattern, Dna):\n",
    "    '''\n",
    "    Distance Between Pattern and String.\n",
    "    Finds the minimum distance between pattern and each string within the Dna collection.\n",
    "    \n",
    "    Args:\n",
    "        Pattern (string): Pattern to search for.\n",
    "        Dna (list[string]): Collection of strings to search for pattern within.\n",
    "        \n",
    "    Returns:\n",
    "        int: Minimum distance between pattern and all strings within the collection.\n",
    "    '''\n",
    "    k = len(Pattern)\n",
    "    dist = 0\n",
    "    for string in Dna:\n",
    "        hamDist = float('inf')\n",
    "        for i in range(len(string) - k + 1):\n",
    "            tmp = hammDist(Pattern, string[i:i+k])\n",
    "            if hamDist > tmp:\n",
    "                hamDist = tmp\n",
    "        dist += hamDist\n",
    "    return dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "# test/run distPatString function\n",
    "\n",
    "pat = 'AAA'\n",
    "Dna = ['TTACCTTAAC', 'GATATCTGTC', 'ACGGCGTTCG', 'CCCTAAAGAG', 'CGTCAGAGGT']\n",
    "print(distPatString(pat, Dna))\n",
    "\n",
    "# with open(\"./Data/dataset_5164_1.txt\") as inFile:\n",
    "#     data = inFile.readlines()\n",
    "\n",
    "# Dna = data[1].strip().split(\" \")\n",
    "# print(distPatString(data[0].strip(), Dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def medString(Dna, k, candidates=False):\n",
    "    '''\n",
    "    Median String Problem.\n",
    "    Find a motif of length k, that has the smallest distance between all of the strings in collection Dna.\n",
    "    \n",
    "    Args:\n",
    "        Dna (list[string]): Collection of DNA sequnces.\n",
    "        k (int): Length of motif to find.\n",
    "        candidates (bool, optional): Return all candidates.\n",
    "            Defualts to false, and returns only the first smallest string.\n",
    "            If true and there are multiple minimum strings returns list of all candidate strings.\n",
    "            \n",
    "    Returns:\n",
    "        string: Sequence that is of length k and has the smallest distance within all sequneces of the collection.\n",
    "    '''\n",
    "    dist = float('inf')\n",
    "    cans = []\n",
    "    for i in range((4 ** k) - 1):\n",
    "        pat = numToPat(i, k)\n",
    "        cDist = distPatString(pat, Dna)\n",
    "        if dist > cDist:\n",
    "            dist = cDist\n",
    "            med = pat\n",
    "            cans = [med]\n",
    "        elif dist == cDist:\n",
    "            cans.append(pat)\n",
    "    if candidates and len(cans) > 1:\n",
    "        return cans\n",
    "    else:\n",
    "        return med"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAC\n"
     ]
    }
   ],
   "source": [
    "# test/run medString function\n",
    "\n",
    "k = 3\n",
    "Dna = ['AAATTGACGCAT', 'GACGACCACGTT', 'CGTCAGCGCCTG', 'GCTGAGCACCGG', 'AGTTCGGGACAG']\n",
    "print(medString(Dna, k))\n",
    "\n",
    "# with open(\"./Data/dataset_158_9.txt\") as inFile:\n",
    "#     data = inFile.readlines()\n",
    "\n",
    "# k = int(data[0].strip())\n",
    "# dna = []\n",
    "# for i in range(1, len(data)):\n",
    "#     dna.append(data[i].strip())\n",
    "# print(medString(dna,k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The median string function works well enough, yet still has a runtime of $O(4^k \\times n \\times k \\times t)$ for larger values of k this solution may be too slow.\n",
    "\n",
    "**Greedy Algorithms**\n",
    "\n",
    "Chooses most attractive alternative at each iteration, optimization occurs over each iteration not globally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profileMost(Text, k, Profile):\n",
    "    '''\n",
    "    Profile-Most Probable k-mer Problem.\n",
    "    Finds k-mer most likely to be generated from the given profile within Text.\n",
    "    \n",
    "    Args:\n",
    "        Text (string): Text to search through.\n",
    "        k (int): Length of target motif.\n",
    "        Profile (list[float]): Nucleotide probability profile for each position in k.\n",
    "        \n",
    "    Returns:\n",
    "        string: Profile most probable k-mer within Text.\n",
    "    '''\n",
    "    profMax = 0.0\n",
    "    kmerMax = Text[:k]\n",
    "    for i in range(1, len(Text) - k + 1):\n",
    "        kmer = Text[i:i+k]\n",
    "        tProf = 1.0\n",
    "        for x in range(k):\n",
    "            tProf *= Profile[symToNum(kmer[x])][x]\n",
    "        if tProf > profMax:\n",
    "            profMax = tProf\n",
    "            kmerMax = kmer\n",
    "    return kmerMax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGAG\n"
     ]
    }
   ],
   "source": [
    "# test/run profile function\n",
    "\n",
    "prof = [[0.2, 0.2, 0.3, 0.2, 0.3], [0.4, 0.3, 0.1, 0.5, 0.1], [0.3, 0.3, 0.5, 0.2, 0.4], [0.1, 0.2, 0.1, 0.1, 0.2]]\n",
    "print(profileMost(\"ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\", 5, prof))\n",
    "\n",
    "# with open(\"./Data/dataset_159_3.txt\") as inFile:\n",
    "#     data = inFile.readlines()\n",
    "\n",
    "# text = data[0].strip()\n",
    "# k = int(data[1].strip())\n",
    "# prof = []\n",
    "# for i in range(2,len(data)):\n",
    "#     prof.append([float(x) for x in data[i].strip().split(\" \")])\n",
    "\n",
    "# print(profile(text, k, prof))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Greedy Motif Search\n",
    "\n",
    "def makeProfile(Dna, pCounts=False):\n",
    "    '''\n",
    "    Generates a nucleotide probability profile for a given collection of same length sequences.\n",
    "    \n",
    "    Args:\n",
    "        Dna (list[string]): Collection of sequences to use for probability calculations\n",
    "        pCounts (bool, optional): Include pseudocounts.\n",
    "            Defaults to False, allows zero probabilities.\n",
    "            If true pseudocounts are introduced according to Laplace's rule of succession.\n",
    "            \n",
    "    Returns:\n",
    "        list[float]: Nucleotide probability profile for given collection of sequences.\n",
    "    '''\n",
    "    prof = [[0] * len(Dna[0]) for _ in range(4)]\n",
    "    for string in Dna:\n",
    "        for i in range(len(string)):\n",
    "            prof[symToNum(string[i])][i] += 1\n",
    "    if pCounts:\n",
    "        return [[(x+1)/(len(Dna) + 4) for x in i] for i in prof]\n",
    "    else:\n",
    "        return [[x/len(Dna) for x in i] for i in prof]\n",
    "\n",
    "def makeCon(prof):\n",
    "    '''\n",
    "    Generate Consensus sequence from a given nucleotide probability profile.\n",
    "    \n",
    "    Args:\n",
    "        prof (list[float]): Nucleotide probability profile.\n",
    "    \n",
    "    Returns:\n",
    "        string: Consensus sequence for the given profile.\n",
    "    '''\n",
    "    con = ''\n",
    "    test = list(zip(*prof))\n",
    "    for pos in test:\n",
    "        con += numToSym(pos.index(max(pos)))\n",
    "    return con\n",
    "\n",
    "def score(motifs, pCounts=False):\n",
    "    '''\n",
    "    Generate a score for a set of motifs.\n",
    "        Score is the summation of the distance between each motif and the consesnus sequence for a given set.\n",
    "        \n",
    "    Args:\n",
    "        motifs (list[string]): Collection of sequences to score.\n",
    "        pCounts (bool, optional): Include pseudocounts.\n",
    "            Defaults to False, allows zero probabilities.\n",
    "            If true pseudocounts are introduced according to Laplace's rule of succession.\n",
    "        \n",
    "    Returns:\n",
    "        int: Score of the motif collection.\n",
    "    '''\n",
    "    prof = makeProfile(motifs, pCounts)\n",
    "    con = makeCon(prof)\n",
    "    score = [hammDist(con, x) for x in motifs]\n",
    "    return sum(score)\n",
    "\n",
    "def greedyMotif(Dna, k, t, pCounts=False):\n",
    "    '''\n",
    "    Greedy Motif Finding Algorithm.\n",
    "    Find a set of motifs that minimizes profile most probable string within each string of the collection.\n",
    "    \n",
    "    Args:\n",
    "        Dna (list[string]): Collection of DNA sequences.\n",
    "        k (int): Length of desired motif.\n",
    "        t (int): Number of sequences.\n",
    "        pCounts (bool, optional): Include pseudocounts.\n",
    "            Defaults to False, allows zero probabilities.\n",
    "            If true pseudocounts are introduced according to Laplace's rule of succession.\n",
    "    \n",
    "    Returns:\n",
    "        list[string]: Collection of motifs that have best score among the collection of sequences.\n",
    "    '''\n",
    "    bestMotifs = [x[0:k] for x in Dna]\n",
    "    bestScore = score(bestMotifs, pCounts)\n",
    "    for i in range(1, len(Dna[0]) - k + 1):\n",
    "        tMotif = [Dna[0][i:i+k]]\n",
    "        for x in range(1,t):\n",
    "            prof = makeProfile(tMotif, pCounts)\n",
    "            tMotif.append(profileMost(Dna[x], k, prof))\n",
    "        tScore = score(tMotif, pCounts)\n",
    "        if tScore < bestScore:\n",
    "            bestMotifs = tMotif\n",
    "            bestScore = tScore\n",
    "    return bestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTC ATC TTC ATC TTC\n"
     ]
    }
   ],
   "source": [
    "# test/run greedyMotif function\n",
    "\n",
    "dna = ['GGCGTTCAGGCA', 'AAGAATCAGTCA', 'CAAGGAGTTCGC', 'CACGTCAATCAC', 'CAATAATATTCG']\n",
    "print(*greedyMotif(dna, 3, 5, True), sep=\" \")\n",
    "\n",
    "# with open(\"./Data/dataset_159_5.txt\") as inFile:\n",
    "# with open(\"./Data/dataset_160_9.txt\") as inFile:\n",
    "#     data = inFile.readlines()\n",
    "    \n",
    "# args = data[0].strip().split(\" \")\n",
    "# k = int(args[0])\n",
    "# t = int(args[1])\n",
    "# dna = []\n",
    "# for i in range(1, len(data)):\n",
    "#     dna.append(data[i].strip())\n",
    "# print(f\"K: {k}\\nT: {t}\\n\")\n",
    "# print(*greedyMotif(dna, k, t, True), sep=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AATCCTA', 'GAACCAC', 'GTAGGAA', 'TAGTTTC']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'AATCCTA'"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Week 3 Quiz\n",
    "test = ['CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC', 'GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC', 'GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG']\n",
    "medString(test, 7, candidates=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
